### ISSUE TITLE
Load a model trained to predict next character or subword instead of next word

### ISSUE THREAD

**Issue by mathetes87** [IndicoDataSolutions/finetune#125]
Hi,

Character based language modeling has its [advantages over word level prediction](https://stats.stackexchange.com/questions/216000/advantage-of-character-based-language-models-over-word-based), and I'm wondering if I'll be able to use this wrapper or not.

My plan is to train a model using Google's T2T as documented [here](https://github.com/tensorflow/tensor2tensor). The model can be trained using subword encoding (default), character or word level encodings. If I were to use any of this options, would the model saved work out of the box with `finetune`? Should I beware of any details when training the model?

The repo looks like it is very well made, I hope this would be seamless. Does anyone know?

**Comment by mathetes87** [comment]
Thanks a lot for the extensive response!

> This model uses byte-pair encoding, which falls somewhere between character based language modeling and traditional word / token based language modeling.

Gotcha

> The tensor2tensor seems to be designed mostly for training models from scratch

Sorry for being vague before, what I actually want to do is train a transformer model in spanish and then do some classification tasks.

> their default recommendations (namely, the use of a transformer model) align with what's already used by default in finetune

That statement left me a little puzzled. Do you mean their model is similar to what finetune expects but not exactly the same?

> you could try to finetune on an unsupervised objective on a large corpus and then swap over to a supervised loss to solve your end task

Given that it is a totally different language, can I do as you say? Or where you assuming that I needed a model for the english language?

Again, thanks a lot for your help

**Comment by madisonmay** [comment]
Thanks for providing more specifics, this is super helpful! 

There may be a way you could use a model trained in tensor2tensor in finetune but it would be a very high effort endeavor and I would be worried that a small difference in implementation could end up causing major problems that might be hard to diagnose. 

I was definitely assuming you needed an English language model (sorry!).  Given that you're looking for a spanish language model, you will probably need to train a new base model from scratch, and finetune is currently missing some of the components you would need to make this work (a method for re-fitting the byte-pair encoder and proper weight initialization).

Given all of this, your best bet might be to go with something like [ULMFit from FastAI](https://github.com/fastai/fastai/tree/master/courses/dl2/imdb_scripts).  It's a bit different (based on LSTM models instead of Transformer models), but the base idea is the same (finetuning language models for classification tasks) and has better support for training from scratch.  There has been [quite a lot of work already](http://forums.fast.ai/t/language-model-zoo-gorilla/14623/3) to make ULMFit work well for languages other than English, including [Spanish in particular](https://github.com/lesscomfortable/SpanishULMFit). 

We intend to add this functionality to `finetune` in the future but there's a fair amount of work to put in before that's possible unfortunately.

**Comment by benleetownsend** [comment]
@mathetes87 Im going to close this for now, if you require further assistance feel free to reopen.
